### PR TITLE
Feature/issue 2471 Move to more recent otto 

### LINKED ISSUE #2471: Move to more recent otto
In particular users have asked for access to JSON.stringify, which was added to otto after our pinned commit (added in otto commit 5fe23327c961b5cdadd65df230f8c2137d8a27e4).

### PR DESCRIPTION
**Pull request by ajres**
fixes #2471 

DO NOT MERGE, otto dependency has been updated to latest otto commit in couchbasedeps repo.

Currently there are build and unit test failures.

### CODE REVIEW

**Comment by adamcfraser** [comment]
@ajres While you wait for the couchbasedeps repo to be updated, can you just refresh your local otto to the latest to check that disabled unit test?

**Comment by ajres** [comment]
Updated local repo from origin to commit a813c59b1b4471ff7ecd3b533bac2f7e7d178784

Updated manifest and re-enabled failing test, the full suite of unit tests pass without any failures.

I'll revert the changes I made to get the build to work on the previous commit.

**Comment by ajres** [comment]
I reverted the code changes and the build completes again.

It looks like the latest commit from origin is a drop in replacement.

**Comment by ajres** [comment]
Waiting for coucbasedeps fork to be updated to the latest origin commit before this PR can be completed.

**Comment by ajres** [comment]
The couchbasedeps repo has been updated to latest commit HASH from origin.

Updated PR to reflect that change and reverted fixes needed for older commit.

**Comment by ajres** [comment]
added otto transient dependency for go-sourcemap/sourcemap.v1

NOTE: Dependency from github.com not couchbasedeps fork, probably need to create a forked repo before merge to master.

**Comment by ajres** [comment]
Currently there are unit test failures, currently investigating root cause and best resolution.

File: db/design_doc.go [hunk 1]
```diff
@@ -244,7 +244,7 @@ func filterViewResult(input sgbucket.ViewResult, user auth.User, applyChannelFil
 			value, ok := row.Value.([]interface{})
 			if ok {
 				// value[0] is the array of channels; value[1] is the actual value
-				if !hasStarChannel || channelsIntersect(visibleChannels, value[0].([]interface{})) {
+				if !hasStarChannel || channelsIntersect(visibleChannels, value[0].([]string)) {
```
Comment (adamcfraser) [inline db/design_doc.go:247]: How is this related to otto?

**Comment by tleyden** [comment]
Since this seems like a fairly risky PR, can this be gated on passing the integration tests?

I kicked off one in progress:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/59/console

Not sure if this will need to get rebased on master though in order to build with go 1.9.2.    However that jenkins job now supports multiple version of go:

* 1.7.6
* 1.9.2
* 1.9

so if this can't get rebased on master, then maybe it can get kicked off again using the appropriate version of go

**Comment by tleyden** [comment]
The jenkins integration job failed.  I tried rebasing on master:

feature/issue_2471_rebased

and kicking off a new job:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/61

**Comment by tleyden** [comment]
It failed with:

```
--- FAIL: TestUserViewQuery (6.48s)
2017-10-30T15:19:30.953-07:00 Feed+: Closing DCP Feed based on termination notification
panic: interface conversion: interface {} is []interface {}, not []string [recovered]
	panic: interface conversion: interface {} is []interface {}, not []string

goroutine 7936 [running]:
testing.tRunner.func1(0xc420a36d20)
	/root/.gvm/gos/go1.9.2/src/testing/testing.go:711 +0x2d2
panic(0xbee6c0, 0xc42041e980)
	/root/.gvm/gos/go1.9.2/src/runtime/panic.go:491 +0x283
github.com/couchbase/sync_gateway/db.filterViewResult(0x2, 0xc420623930, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x11842c0, 0xc420cfc300, ...)
	/var/jenkins/workspace/sync-gateway-integration-master/godeps/src/github.com/couchbase/sync_gateway/db/design_doc.go:226 +0x6e8
github.com/couchbase/sync_gateway/db.(*Database).QueryDesignDoc(0xc420af89e0, 0xceb2a1, 0x3, 0xceb2ab, 0x3, 0xc4206540f0, 0x3, 0xc420f7a4d0, 0x3)
```

I have a feeling this is a known issue though, it sounds familiar.

**Comment by tleyden** [comment]
The `TestUserViewQuery` test failure is mentioned in https://github.com/couchbase/sync_gateway/issues/1580#issuecomment-182729570

File: db/design_doc.go (hunk 1 above)
Comment (adamcfraser) [reply db/design_doc.go:247]: If the underlying issue is the way walrus uses otto for map function execution, should this be fixed there?

Comment (ajres) [reply db/design_doc.go:247]: The unit tests are using wulrus, in walrus views.go in _compileDesignDoc it calls back on:

sgbucket.NewJSMapFunction

Which internally is using otto.

I'm trying to understand if this is strictly a walrus issue or an SG issue.

Comment (ajres) [reply db/design_doc.go:247]: Looking at Traun's comments below for integration test, then it looks like this should be treated as a walrus issue. I'll revert the PR code change and rerun the integration tests.

**Comment by ajres** [comment]
Closing this PR, in favour of #3020 which is based off branch rebased on master
